GitHubBranches.ps1

```diff
@@ -154,6 +155,550 @@ filter Get-GitHubRepositoryBranch
+        [int] $RequiredApprovingReviewCount,
+
+        [string[]] $RestrictPushUsers,
+
+        [string[]] $RestrictPushTeams,
+
+        [string[]] $RestrictPushApps,
+
+        [switch] $RequireLinearHistory,
+
+        [switch] $AllowForcePushes,
+
+        [switch] $AllowDeletions,
+
+        [string] $AccessToken,
+
+        [switch] $NoStatus
+    )
+
+    $elements = Resolve-RepositoryElements
+    $OwnerName = $elements.ownerName
+    $RepositoryName = $elements.repositoryName
+
+    $telemetryProperties = @{
+        OwnerName = (Get-PiiSafeString -PlainText $OwnerName)
+        RepositoryName = (Get-PiiSafeString -PlainText $RepositoryName)
+    }
+
+    if ($PSBoundParameters.ContainsKey('StatusChecks'))
+    {
+        $requiredStatusChecks = @{
+            strict = $RequireUpToDateBranches.ToBool()
+            contexts = $StatusChecks
+        }
+    }
+    else
+    {
+        $requiredStatusChecks = $null
+    }
+
+    $dismissalRestrictions = @{}
+
+    if ($PSBoundParameters.ContainsKey('DismissalUsers'))
+    {
+       $dismissalRestrictions['users'] = $DismissalUsers
+    }
+    if ($PSBoundParameters.ContainsKey('DismissalTeams'))
+    {
+        $teams = Get-GitHubTeam -OwnerName $OwnerName -RepositoryName $RepositoryName |
+            Where-Object -FilterScript { $DismissalTeams -contains $_.name }
+        $dismissalRestrictions['teams'] = @($teams.slug)
+    }
+
+    $requiredPullRequestReviews = @{}
+
+    if ($PSBoundParameters.ContainsKey('DismissStaleReviews'))
+    {
+        $requiredPullRequestReviews['dismiss_stale_reviews'] = $DismissStaleReviews.ToBool()
+    }
+    if ($PSBoundParameters.ContainsKey('RequireCodeOwnerReviews'))
+    {
+        $requiredPullRequestReviews['require_code_owner_reviews'] = $RequireCodeOwnerReviews.ToBool()
+    }
+    if ($PSBoundParameters.ContainsKey('RequiredApprovingReviewCount'))
+    {
+        $requiredPullRequestReviews['required_approving_review_count'] = $RequiredApprovingReviewCount
+    }
+
+    if ($dismissalRestrictions.count -gt 0)
+    {
+        $requiredPullRequestReviews['dismissal_restrictions'] = $dismissalRestrictions
+    }
+
+    if ($requiredPullRequestReviews.count -eq 0)
+    {
+        $requiredPullRequestReviews = $null
+    }
+
+    if ($PSBoundParameters.ContainsKey('RestrictPushUsers') -or
+        $PSBoundParameters.ContainsKey('RestrictPushTeams') -or
+        $PSBoundParameters.ContainsKey('RestrictPushApps'))
+    {
+        if ($null -eq $RestrictPushUsers)
+        {
+            $RestrictPushUsers = @()
+        }
+
+        if ($null -eq $RestrictPushTeams)
+        {
+            $restrictPushTeamSlugs = @()
+        }
+        else
+        {
+            $teams = Get-GitHubTeam -OwnerName $OwnerName -RepositoryName $RepositoryName |
+                Where-Object -FilterScript { $RestrictPushTeams -contains $_.name }
+            $restrictPushTeamSlugs['teams'] = @($teams.slug)
+        }
+
+        $restrictions = @{
```

Comment:
Let's continue one of the remaining [unresolved conversations](https://github.com/microsoft/PowerShellForGitHub/pull/238/files/2217645bf05363fa52fc35bf91bdbebc84f8204c#r443135683) from the other PR in this one.

Given that you have implemented `Get-GitHubRepositorBranchProtectionRule`, it seems that you should be able to query its _current_ state first, and then fill in the data for any required property that the user hasn't supplied so that you don't unintentionally wipe out existing rules that they have in place.

Reply:
I've decided to change the function from `Set` to `New` to follow the PowerShell cmdlet pattern of New/Get/Set/Remove. A `Set` function can be added in another PR.

Reply:
Works for me.  That will definitely simplify this.